kubectl-plugin/pkg/cmd/submit/submit.go

```diff
@@ -0,0 +1,486 @@
+		time.Sleep(2 * time.Second)
+		fmt.Printf("Checking Cluster Status for cluster %s...\n", options.cluster)
+		currCluster, err := k8sClients.DynamicClient().Resource(util.RayClusterGVR).Namespace(*options.configFlags.Namespace).Get(ctx, options.cluster, v1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("Failed to get cluster information with error: %w", err)
+		}
+		ready, err = extractClusterStatus(currCluster)
+		if err != nil {
+			return fmt.Errorf("Failed to get cluster status: %w", err)
+		}
+		currTime = time.Now()
+	}
+
+	if !ready {
+		if options.RayJob != nil {
+			fmt.Printf("Deleting RayJob...\n")
+			err = k8sClients.DynamicClient().Resource(util.RayJobGVR).Namespace(*options.configFlags.Namespace).Delete(ctx, options.RayJob.GetName(), v1.DeleteOptions{})
+			if err != nil {
+				return fmt.Errorf("Failed to clean up ray job after time out.: %w", err)
+			}
+			fmt.Printf("Cleaned Up RayJob: %s\n", options.RayJob.GetName())
+		}
+		return fmt.Errorf("Timed out waiting for cluster")
+	}
+
+	svcName, err := k8sClients.GetRayHeadSvcName(ctx, *options.configFlags.Namespace, util.RayCluster, options.cluster)
+	if err != nil {
+		return fmt.Errorf("Failed to find service name: %w", err)
+	}
+
+	// Only start port forward if address is not inputted, if there is already an address that means there is already a way to reach the raycluster
+	if len(options.address) <= 0 {
+		portForwardCmd := portforward.NewCmdPortForward(factory, *options.ioStreams)
+		portForwardCmd.SetArgs([]string{"service/" + svcName, fmt.Sprintf("%d:%d", 8265, 8265)})
+
+		options.address = "http://localhost:8265"
+
+		// create new context for port-forwarding so we can cancel the context to stop the port forwarding only
+		portforwardctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		go func() {
+			fmt.Printf("Port Fowarding service %s\n", svcName)
+			if err := portForwardCmd.ExecuteContext(portforwardctx); err != nil {
+				log.Fatalf("Error occurred while port-forwarding Ray dashboard: %v", err)
+			}
+		}()
+	}
+
+	raySubmitCmd, err := options.raySubmitCmd()
+	if err != nil {
+		return fmt.Errorf("failed to create Ray submit command with error: %w", err)
+	}
+	fmt.Printf("Ray command: %v\n", raySubmitCmd)
+	cmd := exec.Command(raySubmitCmd[0], raySubmitCmd[1:]...) //nolint:gosec // command is sanitized in raySubmitCmd() and file paths are cleaned in Complete()
+
+	// Get the outputs/pipes for `ray job submit` outputs
+	rayCmdStdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("Error while setting up `ray job submit` stdout: %w", err)
+	}
+	rayCmdStdErr, err := cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("Error while setting up `ray job submit` stderr: %w", err)
+	}
+
+	go func() {
+		fmt.Printf("Running ray submit job command...\n")
+		err := cmd.Start()
+		if err != nil {
+			log.Fatalf("error occurred while running command %s: %v", fmt.Sprint(raySubmitCmd), err)
+		}
+	}()
```

Comment:
There is a race condition here. We should wait until port 8265 is ready.

![image](https://github.com/user-attachments/assets/d4a515dd-882b-485c-a32f-1e2e6edfbe8c)

Reply:
Since the kubectl port foward doesn't provide a way of polling the status or status checking mechanism, I created a new pipe for the output and checked for the `Forwarding from` string. 

The other solutions that I also thought about was have a timeout and assume that either the portforward succeeded  or it failed, in which case ends the function. Or we check stdout for the `Forwarding from` string.

Reply:
I think another solution would be trying to make a connection to port 8265. Much like what a readiness probe in Kubernetes does.

Reply:
Hmm, okay I will try this.

Reply:
I changed how we are waiting for the port forward. Please take a look.